### PR TITLE
Change make_quadratic to use a loop rather than recursion

### LINKED ISSUE #556: Exceeded maximum recursion depth when converting large BinaryPolynomial to QUBO
When a large `BinaryPolynomial` instance is converted to a QUBO using `make_quadratic`, the recursive procedure exceeds maximum recursion depth.

Here is an example that reproduces the problem.
```
import dimod
from itertools import combinations
terms = combinations(range(20), 15)
poly = {k: 1.0 for k in terms}
bqm = dimod.make_quadratic(poly, 1.0, 'SPIN')
```

This should successfully return a bqm object instance. However, an error is raised instead,
`RecursionError: maximum recursion depth exceeded while calling a Python object`
and,
`TypeError: unsupported callable`

**Environment**
 - OS: [e.g., Ubuntu 18.04.4 LTS]
 - Python version: [e.g., 3.7.3]
 - Default maximum recursion depth limit

**Related issues**
This may be related to #550

### PR DESCRIPTION
**Pull request by arcondello**
Closes #556

### CODE REVIEW

File: dimod/higherorder/utils.py [hunk 1]
```diff
@@ -134,87 +152,77 @@ def make_quadratic(poly, strength, vartype=None, bqm=None):
     # we want to be able to mutate the polynomial so copy. We treat this as a
     # dict but by using BinaryPolynomail we also get automatic handling of
```
Comment (conta877) [inline dimod/higherorder/utils.py]: typo: BinaryPolynomail -> BinaryPolynomial

File: dimod/higherorder/utils.py [hunk 2]
```diff
@@ -134,87 +152,77 @@ def make_quadratic(poly, strength, vartype=None, bqm=None):
...
+            u, v = term
+            bqm.add_interaction(u, v, bias)
+        elif len(term) == 1:
+            v, = term
+            bqm.add_variable(v, bias)
+        elif len(term) == 0:
+            bqm.add_offset(bias)
+        else:
+            # still has higher order terms, this shouldn't happen
+            raise RuntimeError
```
Comment (conta877) [inline dimod/higherorder/utils.py]: can we add a better error message here.

Comment (arcondello) [reply dimod/higherorder/utils.py]: Hmm, this is an "internal error", but suppose I could say that the user should file a bug report.

Comment (conta877) [reply dimod/higherorder/utils.py]: yeah it is internal - but if it happens its a legit bug and will help with bug reporting.

Comment (conta877) [reply dimod/higherorder/utils.py]: Also will save frustration for users.

**Pull request merged**